### ISSUE TITLE
Multiline parsers doesn't concatenate structured logs

### ISSUE THREAD

**Issue by sophieyfang** [fluent/fluent-bit#4554]
## Bug Report

**Describe the bug**
Multiline parsers doesn't concatenate structured logs

**To Reproduce**
- configuration file:
 ```
sophieyfang_google_com@debian10-meow:~$ cat fluent-bit-json.conf 
[SERVICE]
    flush                 1
    log_level             info
    parsers_file          parsers_multiline_json.conf

[INPUT]
    name                  tail
    path                  json_string.log
    read_from_head        true
    tag                   wow

[FILTER]
    Key_Name log
    Match    *
    Name     parser
    Parser   parse_json_0

[FILTER]
    name                  multiline
    match                 *
    multiline.parser      multiline-regex-test

[OUTPUT]
    name                  stdout
    match                 *
    
sophieyfang_google_com@debian10-meow:~$ cat parsers_multiline_json.conf 
[PARSER]
    Name        parse_json_0
    Format      json
[MULTILINE_PARSER]
    name          multiline-regex-test
    key_content   message
    type          regex
    flush_timeout 1000
    #
    # Regex rules for multiline parsing
    # ---------------------------------
    #
    # configuration hints:
    #
    #  - first state always has the name: start_state
    #  - every field in the rule must be inside double quotes
    #
    # rules |   state name  | regex pattern                  | next state
    # ------|---------------|--------------------------------------------
    rule      "start_state"   "/(Dec \d+ \d+\:\d+\:\d+)(.*)/"  "cont"
    rule      "cont"          "/^\s+at.*/"                     "cont"
 ```
- Example log message:
```
sophieyfang_google_com@debian10-meow:~$ cat json_string.log 
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":"Dec 14 06:41:08 Exception in thread main java.lang.RuntimeException: Something has gone wrong, aborting!"}
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":"at com.myproject.module.MyProject.badMethod(MyProject.java:22)"}
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":"at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:18)"}
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":"at com.myproject.module.MyProject.anotherMethod(MyProject.java:14)"}
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":"at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:13)"}
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":"at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:12)"}
```
- Steps to reproduce the problem:
```
sophieyfang_google_com@debian10-meow:~$ /opt/google-cloud-ops-agent/subagents/fluent-bit/bin/fluent-bit --config fluent-bit-json.conf 
Fluent Bit v1.8.11
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2021/12/29 19:35:15] [ info] [engine] started (pid=5657)
[2021/12/29 19:35:15] [ info] [storage] version=1.1.5, initializing...
[2021/12/29 19:35:15] [ info] [storage] in-memory
[2021/12/29 19:35:15] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2021/12/29 19:35:15] [ info] [cmetrics] version=0.2.2
[2021/12/29 19:35:15] [ info] [input:tail:tail.0] inotify_fs_add(): inode=143355 watch_fd=1 name=json_string.log
[0] wow: [1640806515.466821930, {"key1"=>12345, "key2"=>"abc", "time"=>"2006-07-28T13:22:04Z", "message"=>"Dec 14 06:41:08 Exception in thread main java.lang.RuntimeException: Something has gone wrong, aborting!"}]
[1] wow: [1640806515.466826081, {"key1"=>12345, "key2"=>"abc", "time"=>"2006-07-28T13:22:04Z", "message"=>"at com.myproject.module.MyProject.badMethod(MyProject.java:22)"}]
[2] wow: [1640806515.466826329, {"key1"=>12345, "key2"=>"abc", "time"=>"2006-07-28T13:22:04Z", "message"=>"at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:18)"}]
[3] wow: [1640806515.466826493, {"key1"=>12345, "key2"=>"abc", "time"=>"2006-07-28T13:22:04Z", "message"=>"at com.myproject.module.MyProject.anotherMethod(MyProject.java:14)"}]
[4] wow: [1640806515.466826637, {"key1"=>12345, "key2"=>"abc", "time"=>"2006-07-28T13:22:04Z", "message"=>"at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:13)"}]
[5] wow: [1640806515.466826827, {"key1"=>12345, "key2"=>"abc", "time"=>"2006-07-28T13:22:04Z", "message"=>"at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:12)"}]
^C[2021/12/29 19:35:30] [engine] caught signal (SIGINT)
[2021/12/29 19:35:30] [ info] [input] pausing tail.0
[2021/12/29 19:35:30] [ warn] [engine] service will shutdown in max 5 seconds
[2021/12/29 19:35:31] [ info] [engine] service has stopped (0 pending tasks)
[2021/12/29 19:35:31] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=143355 watch_fd=1
```
**Expected behavior**
Instead of generating one line each as shown above, it should've concatenated them and generated one log entry..

**Your Environment**
* Version used: 1.8.11
```
sophieyfang_google_com@debian10-meow:~$ cat /etc/os-release 
PRETTY_NAME="Debian GNU/Linux 10 (buster)"
NAME="Debian GNU/Linux"
VERSION_ID="10"
VERSION="10 (buster)"
VERSION_CODENAME=buster
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

**Comment by sophieyfang** [comment]
@edsiper Not sure if i use the configuration file right or it might be a real issue 😊

**Comment by edsiper** [comment]
Thanks for reporting this issue.

I found two issues associated:

__1. Use case needs to be adjusted__

Note that the continuation regular expression expects a space character at the beginning (__\s__):

```
"/^\s+at.*/" 
```

So the JSON logs should look like this:

```json
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.anotherMethod(MyProject.java:14)"}
```

> note the initial empty space on ` at`

__2. Fluent BIt Bug__

Tail plugin invokes Multiline API properly, but the flush [callback defined](https://github.com/fluent/fluent-bit/blob/master/plugins/in_tail/tail_file.c#L783-L795) ingest the data into the `file->mult_sbuf` msgpack buffer instead of writing it directly into the engine.

I have already a fix for the issue mentioned, but also I found that when a line that does not match the expected pattern is ingested, the key used to package it is the same one configured in the multiline parser `message`, the expectation is that must be `log` (still thinking about this).

Example input logs:

```
happy new
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":"Dec 14 06:41:08 Exception in thread main java.lang.RuntimeException: Something has gone wrong, aborting!"}
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.badMethod(MyProject.java:22)"}
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:18)"}
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.anotherMethod(MyProject.java:14)"}
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:13)"}
{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:12)"}
year
```

Current output with WIP patch:

```
[0] wow: [1640997871.543141190, {"message"=>"happy new"}]
[1] wow: [1640997871.543169806, {"key1"=>12345, "key2"=>"abc", "time"=>"2006-07-28T13:22:04Z", "message"=>"Dec 14 06:41:08 Exception in thread main java.lang.RuntimeException: Something has gone wrong, aborting!
 at com.myproject.module.MyProject.badMethod(MyProject.java:22)
 at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:18)
 at com.myproject.module.MyProject.anotherMethod(MyProject.java:14)
 at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:13)
 at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:12)"}]
[2] wow: [1640997871.543169806, {"message"=>"year"}]
```

PR will arrive early next year :)

**Comment by sophieyfang** [comment]
Thanks Eduardo! Also Happy New Year! 😄
Yes, the space front missing from the log was the problem. Good catch! 

I seem to find a new problem:

The `parser` property of multiline parser seems not working properly, description of this parser property:
```
Name of a pre-defined parser that must be applied to the incoming content before applying the regex rule. If no parser is defined, it's assumed that's a raw text and not a structured message. 

Note: when a parser is applied to a raw text, then the regex is applied against a specific key of the structured message by using the  key_content  configuration property (see below).
```

### Configuration files:
```
sophieyfang_google_com@debian10-meow:~$ cat fluent-bit-json.conf 
[SERVICE]
    flush                 1
    log_level             info
    parsers_file          parsers_multiline_json.conf

[INPUT]
    name                  tail
    path                  json_string.log
    read_from_head        true
    tag                   wow


[FILTER]
    name                  multiline
    match                 *
    multiline.parser      multiline-regex-test

[OUTPUT]
    name                  stdout
    match                 *
    
sophieyfang_google_com@debian10-meow:~$ cat parsers_multiline_json.conf 
[PARSER]
    Name        parse_json_0
    Format      json
[MULTILINE_PARSER]
    name          multiline-regex-test
    key_content   message
    type          regex
    parser        parse_json_0 
    flush_timeout 1000
    #
    # Regex rules for multiline parsing
    # ---------------------------------
    #
    # configuration hints:
    #
    #  - first state always has the name: start_state
    #  - every field in the rule must be inside double quotes
    #
    # rules |   state name  | regex pattern                  | next state
    # ------|---------------|--------------------------------------------
    rule      "start_state"   "/(Dec \d+ \d+\:\d+\:\d+)(.*)/"  "cont"
    rule      "cont"          "/^\s+at.*/"                     "cont"
    
```

### Reproduce
```
sophieyfang_google_com@debian10-meow:~$ /opt/google-cloud-ops-agent/subagents/fluent-bit/bin/fluent-bit --config fluent-bit-json.conf 
Fluent Bit v1.8.11
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/01/01 15:29:08] [ info] [engine] started (pid=32305)
[2022/01/01 15:29:08] [ info] [storage] version=1.1.5, initializing...
[2022/01/01 15:29:08] [ info] [storage] in-memory
[2022/01/01 15:29:08] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2022/01/01 15:29:08] [ info] [cmetrics] version=0.2.2
[2022/01/01 15:29:08] [ info] [input:tail:tail.0] inotify_fs_add(): inode=138350 watch_fd=1 name=json_string.log
[0] wow: [1641050948.505215274, {"log"=>"{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":"Dec 14 06:41:08 Exception in thread main java.lang.RuntimeException: Something has gone wrong, aborting!"}"}]
[1] wow: [1641050948.505218833, {"log"=>"{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.badMethod(MyProject.java:22)"}"}]
[2] wow: [1641050948.505219077, {"log"=>"{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:18)"}"}]
[3] wow: [1641050948.505219227, {"log"=>"{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.anotherMethod(MyProject.java:14)"}"}]
[4] wow: [1641050948.505219353, {"log"=>"{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:13)"}"}]
[5] wow: [1641050948.505219531, {"log"=>"{"key1": 12345, "key2": "abc", "time": "2006-07-28T13:22:04Z","message":" at com.myproject.module.MyProject.oneMoreMethod(MyProject.java:12)"}"}]
^C[2022/01/01 15:29:40] [engine] caught signal (SIGINT)
[2022/01/01 15:29:40] [ info] [input] pausing tail.0
[2022/01/01 15:29:40] [ warn] [engine] service will shutdown in max 5 seconds
[2022/01/01 15:29:41] [ info] [engine] service has stopped (0 pending tasks)
[2022/01/01 15:29:41] [ info] [input:tail:tail.0] inotify_fs_remove(): inode=138350 watch_fd=1
```

### Expected behavior
It shall concatenate into one log entry

**Comment by sophieyfang** [comment]
This actually might be a blocker for us. For the cases, that  client has test.log like below:

```
"_HOSTNAME":"debian10-meow", "_SYSTEMD_SLICE":"system.slice", "MESSAGE":"2021-12-28T09:45:43.264Z        error        scraperhelper/scrapercontroller.go:198        Error scraping metrics        {"kind": "receiver", "name": "hostme”}"
"_HOSTNAME":"debian10-meow", "_SYSTEMD_SLICE":"system.slice", "MESSAGE":"go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport’
"_HOSTNAME":"debian10-meow", "_SYSTEMD_SLICE":"system.slice", "MESSAGE":"	/work/submodules/opentelemetry-operations-collector/vendor/go.opentelemetry.io/collector/receiver/scraperhelper/scrapercontroller.go:198"
"_HOSTNAME":"debian10-meow", "_SYSTEMD_SLICE":"system.slice", "MESSAGE":"go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1"
"_HOSTNAME":"debian10-meow", "_SYSTEMD_SLICE":"system.slice", "MESSAGE":"	/work/submodules/opentelemetry-operations-collector/vendor/go.opentelemetry.io/collector/receiver/scraperhelper/scrapercontroller.go:173"
```

# I have tried two different set of confs:
## Approach 1 -- doesn't work

```
sophieyfang_google_com@debian10-meow:~$ cat parsers_multiline_json.conf 
[PARSER]
    Name        parse_json_0
    Format      json
[MULTILINE_PARSER]
    name          multiline-regex-test
    key_content   message
    type          regex
    parser        parse_json_0 
    flush_timeout 1000
    #
    # Regex rules for multiline parsing
    # ---------------------------------
    #
    # configuration hints:
    #
    #  - first state always has the name: start_state
    #  - every field in the rule must be inside double quotes
    #
    # rules |   state name  | regex pattern                  | next state
    # ------|---------------|--------------------------------------------
    rule      "start_state"   "(\d+-\d+-\d+T\d+:\d+:\d+.\d+Z)(.*)"  "cont"
    rule      "cont"          "go.opentelemetry.io.*"               "cont1"
    rule      "cont1"         "^\s+.*"                              "cont"

sophieyfang_google_com@debian10-meow:~$ cat fluent-bit-json.conf 
[SERVICE]
    flush                 1
    log_level             info
    parsers_file          parsers_multiline_json.conf

[INPUT]
    name                  tail
    path                  json_string.log
    read_from_head        true
    multiline.parser      multiline-regex-test

[OUTPUT]
    name                  stdout
    match                 *
```

## Approach 2 --  doesn't work
```
sophieyfang_google_com@debian10-meow:~$ cat parser.conf 
[PARSER]
    Name        parse_json_0
    Format      json
[MULTILINE_PARSER]
    name          default_pipeline.multiline_parser_1
    type          regex
    flush_timeout 1000
    #
    # Regex rules for multiline parsing
    # ---------------------------------
    #
    # configuration hints:
    #
    #  - first state always has the name: start_state
    #  - every field in the rule must be inside double quotes
    #
    # rules |   state name  | regex pattern                  | next state
    # ------|---------------|--------------------------------------------
    rule      "start_state"   "(\d+-\d+-\d+T\d+:\d+:\d+.\d+Z)(.*)"  "cont"
    rule      "cont"          "go.opentelemetry.io.*"               "cont1"
    rule      "cont1"         "^\s+.*"                              "cont"

sophieyfang_google_com@debian10-meow:~$ cat main.conf
[SERVICE]
    flush        1
    log_level    info
    parsers_file parser.conf 
[INPUT]
    name             tail
    path             systemd.log
    read_from_head   true
    parser           parse_json_0
   multiline.parser go, python, java, default_pipeline.multiline_parser_1
[OUTPUT]
    name             stdout
    match            *
```

**Comment by sophieyfang** [comment]
The issue has been fixed in branch: 1.8
```
[sophieyfang_google_com@centos8-6 ~]$ cat fluent-bit-json.conf 
[SERVICE]
    flush                 1
    log_level             info
    parsers_file          parsers_multiline_json.conf

[INPUT]
    name                  tail
    path                  systemd.log
    read_from_head        true
    multiline.parser      multiline-regex-test

[OUTPUT]
    name                  stdout
    match                 *
[sophieyfang_google_com@centos8-6 ~]$ cat parsers_multiline_json.conf 
[PARSER]
    Name        parse_json_0
    Format      json
[MULTILINE_PARSER]
    name          multiline-regex-test
    key_content   MESSAGE
    type          regex
    parser        parse_json_0 
    flush_timeout 1000
    #
    # Regex rules for multiline parsing
    # ---------------------------------
    #
    # configuration hints:
    #
    #  - first state always has the name: start_state
    #  - every field in the rule must be inside double quotes
    #
    # rules |   state name  | regex pattern                  | next state
    # ------|---------------|--------------------------------------------
    rule      "start_state"   "(\d+-\d+-\d+T\d+:\d+:\d+.\d+Z)(.*)"  "cont"
    rule      "cont"          "go.opentelemetry.io.*"               "cont1"
    rule      "cont1"         "^\s+.*"                              "cont"
[sophieyfang_google_com@centos8-6 ~]$ cat systemd.log 
{"_HOSTNAME":"debian10-meow", "_SYSTEMD_SLICE":"system.slice", "MESSAGE":"2021-12-28T09:45:43.264Z        error        scraperhelper/scrapercontroller.go:198        Error scraping metrics        {\"kind\": \"receiver\", \"name\": \"hostme\"}"}
{"_HOSTNAME":"debian10-meow", "_SYSTEMD_SLICE":"system.slice", "MESSAGE":"go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport"}
{"_HOSTNAME":"debian10-meow", "_SYSTEMD_SLICE":"system.slice", "MESSAGE":"	/work/submodules/opentelemetry-operations-collector/vendor/go.opentelemetry.io/collector/receiver/scraperhelper/scrapercontroller.go:198"}
{"_HOSTNAME":"debian10-meow", "_SYSTEMD_SLICE":"system.slice", "MESSAGE":"go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1"}
{"_HOSTNAME":"debian10-meow", "_SYSTEMD_SLICE":"system.slice", "MESSAGE":"	/work/submodules/opentelemetry-operations-collector/vendor/go.opentelemetry.io/collector/receiver/scraperhelper/scrapercontroller.go:173"}
[sophieyfang_google_com@centos8-6 ~]$ fluent-bit/build/bin/fluent-bit --conf fluent-bit-json.conf 
Fluent Bit v1.8.12
* Copyright (C) 2019-2021 The Fluent Bit Authors
* Copyright (C) 2015-2018 Treasure Data
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/01/06 18:22:25] [ info] [engine] started (pid=111381)
[2022/01/06 18:22:25] [ info] [storage] version=1.1.5, initializing...
[2022/01/06 18:22:25] [ info] [storage] in-memory
[2022/01/06 18:22:25] [ info] [storage] normal synchronization mode, checksum disabled, max_chunks_up=128
[2022/01/06 18:22:25] [ info] [cmetrics] version=0.2.2
[2022/01/06 18:22:25] [ info] [input:tail:tail.0] multiline core started
[2022/01/06 18:22:25] [ info] [sp] stream processor started
[2022/01/06 18:22:25] [ info] [input:tail:tail.0] inotify_fs_add(): inode=51147294 watch_fd=1 name=systemd.log
[0] tail.0: [1641493345.520827649, {"_HOSTNAME"=>"debian10-meow", "_SYSTEMD_SLICE"=>"system.slice", "MESSAGE"=>"2021-12-28T09:45:43.264Z        error        scraperhelper/scrapercontroller.go:198        Error scraping metrics        {"kind": "receiver", "name": "hostme"}
go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).scrapeMetricsAndReport
	/work/submodules/opentelemetry-operations-collector/vendor/go.opentelemetry.io/collector/receiver/scraperhelper/scrapercontroller.go:198
go.opentelemetry.io/collector/receiver/scraperhelper.(*controller).startScraping.func1
	/work/submodules/opentelemetry-operations-collector/vendor/go.opentelemetry.io/collector/receiver/scraperhelper/scrapercontroller.go:173"}]
```
